File: hw2/codeforces/badCommit.py
import sys

cache = dict()

bad = 0
good = int(input())

while good - bad > 1:
    mid = (bad + good) // 2
    answer = cache.get(mid)

    if not answer:
        print(mid)
        sys.stdout.flush()
        answer = int(input())
        cache[mid] = answer

    if answer == 0:
        good = mid
    else:
        bad = mid

print("! {}".format(good))
sys.stdout.flush()
sys.exit()

File: hw2/codeforces/badCommit.js
function flush() {
    process.stdout.clearLine();
    process.stdout.cursorTo(0);
}

function programm() {
    let bad = -1;
    let good = 0;
    let number = 0;

    function testResult(input) {
        flush();
        const value = +input;

        if (value > 1) good = value;
        if (value <= 1) {
            if (value === 0) good = number;
            else bad = number;
        }
        
        if (good - bad <= 1) {
            process.stdin.write(`! ${good}`);
            process.exit(0);
        }
        
        number = parseInt(((bad + good) / 2));

        process.stdin.write(`${number}\n`);
        flush();
    }

    return {
        testResult
    }
}

const p = programm();

process.stdin.resume();
process.stdin.setEncoding("ascii");
process.stdin.on("data", (input) => p.testResult(input));